### ISSUE TITLE
Offline CRUD and sync to server

### ISSUE THREAD

**Issue by johnpitcairn** [signalpoint/DrupalGap#258]
I'm tasked with developing an app for use by remote workers that will wind up running without a data connection some of the time. I'm going to need to allow users to create, update and delete content while they are offline, then sync changes back to Drupal when they have a data connection.

Initial examination of drupalgap suggests there is no local entity data model, so I'd wind up building data models with backbone.js anyway. If so, I might as well go fully offline-first, forget about drupalgap, and handle sync to Drupal myself (possibly Drupal 8).

Or am I missing something in drupalgap that would help me?

**Comment by signalpoint** [comment]
@johnpitcairn Out of the box, DrupalGap does not have this feature. This has been requested a few times, and I know of one person who successfully implemented it. Private message me on d.o. and I can put you in touch with her.

Alternatively, I really would like to turn this into a contrib module for DrupalGap. Essentially, the module just needs to use a hook_form_alter() within the DG FAPI and attach a custom submit handler, that in turns check's for offline, then stores items in local storage, etc, etc.

I wouldn't mind assisting in writing this module, just let me know if you're interested. If you're going to DrupalCon Austin, we could sit down and sprint something together I think.

UPDATE: here is a rough overview of what needs to be implemented for this fix:

# Issue Tasks
- [ ] Use hook_services_\* to save entity field and instance to local storage after the initial system connect call
- [ ] When offline, all forms need a custom submit handler attached (probably pre pended or over write any existing submit handlers). This submit handler must then save the form state values into a local storage queue, along with the original submit handler. (Except user login and registration).
- [ ] When offline, all forms must check to see if any form state values have been queued up for synchronization, if they have been, place them back into the form
- [ ] When the device comes back online, it should run the form state queue and submit each form that was queued up
- [ ] We need same way to detect when a device goes on/off offline, maybe a background service is needed. At minimum, the jDrupal services API needs to check for a connection, and if it doesn't have one, the call should probably be queued up for later

**Comment by johnpitcairn** [comment]
Yes, I'm interested. I won't be at DrupalCon Austin (I live in New Zealand). I can start hacking something together. If the Drupal 8 effort is far enough along I'd also be very interested in working with that.

**Comment by signalpoint** [comment]
@johnpitcairn The D8 version still needs lots of love, it isn't functional at all at this point. I started an 8.x-1.x branch of the Drupal module, but didn't make it very far. Any help is much appreciated!

**Comment by johnpitcairn** [comment]
I think I'll back off on this (for now). A closer analysis indicates I really will need full-featured local entity data models and business logic, otherwise entity creation while offline will not work well (validation, permissions, relationships, etc). You can't simply queue an entity creation form, the user won't be able to work with the result until online again.

So my application needs to be offline-first, possibly with Drupal as the data store and admin-only back-end. Backbone.dualstorage here I come (I think).

**Comment by signalpoint** [comment]
Just to update everyone on this issue:
- it is the most requested feature
- it is very complex to implement
- it needs sponsors and/or contributors
- the successful implementation mentioned earlier never happened, it doesn't exist :P

www.drupalgap.org/contribute

Stop by #drupal-drupalgap on IRC if you're interested in implementing this solution, I have many thoughts rolling around in my head, so it'd be nice to chat with someone about what it takes to accomplish this.

Interesting read: http://coenraets.org/blog/2012/05/simple-offline-data-synchronization-for-mobile-web-and-phonegap-applications/

**Comment by ssoulless** [comment]
I would like to help to accomplish this. let me know

**Comment by signalpoint** [comment]
First we need to decide what storage mechanism(s) will be used: http://docs.phonegap.com/en/3.3.0/cordova_storage_storage.md.html#Storage

Right now DrupalGap uses local storage for everything, and has no api for any of the other mechanisms.

I guess for simplicity, going the local storage route may be appropriate. I don't know enough about all the mechanisms to weight the pros/cons of each.

**Comment by ssoulless** [comment]
in http://coenraets.org/blog/2012/05/simple-offline-data-synchronization-for-mobile-web-and-phonegap-applications/

it only supports unidirectional sync, but in the example there you cannot make changes in the client and update them to the server, that is the feature we need =/

**Comment by ssoulless** [comment]
I began to do some experiments with the code in https://github.com/ccoenraets/offline-sync

maybe there we can have some conclutions... good start point

**Comment by ssoulless** [comment]
Here http://docs.phonegap.com/en/3.3.0/cordova_file_file.md.html#File describes everything how to use local file system, is drupalgap using local file system? 

and Im agreed with you, going the local storage route is appropriate, I gues drupal gap uses WebSQL, does not it?

**Comment by signalpoint** [comment]
@ssoulless Yes, DrupalGap uses Local Storage. DrupalGap does NOT use WebSQL or any other method.

UPDATE: I've made a list of the tasks that need to be implemented to get this feature working: https://github.com/signalpoint/DrupalGap/issues/258#issuecomment-43841685

**Comment by ssoulless** [comment]
Ok are we gonna manage this through a separate drupal module? or are we gonna work on the drupalgap core?

**Comment by signalpoint** [comment]
@ssoulless DrupalGap core will be used for this feature.

**Comment by samwilson** [comment]
Is there any news on this feature? I'd be interested in helping, if I can.

**Comment by signalpoint** [comment]
@samwilson Anyone interested in helping on this issue, please stop by #drupal-drupalgap on IRC to chat about it, thanks!

**Comment by unmeshsheth** [comment]
I am very interested in this feature as this is a single most requested feature in my line of work.  Happy to see lot of work. However, I am confused with above thread! Is someone still working on it or is this issue dead?

**Comment by johnpitcairn** [comment]
I am not working on it.

On 27 February 2015 at 11:37, Unmesh Sheth notifications@github.com wrote:

> I am very interested in this feature as this is a single most requested
> feature in my line of work. Happy to see lot of work. However, I am
> confused with above thread! Is someone still working on it or is this issue
> dead?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/signalpoint/DrupalGap/issues/258#issuecomment-76290244
> .

## 

_John Pitcairn_
_Opus Locus_
Auckland, New Zealand
Tel: +64 9 815 5984
Mob: +64 27 474 8636

**Comment by hounded** [comment]
Where is this up to ? 

I'm trying to build a phonegap app that can save form data when there is no connectivity and then create new nodes when connectivity is achieved. 

I was looking at angularjs and localforage to do this, but as the ultimate goal is to integrate with a drupal website this looks like the logical answer. 

What I have described, is this achievable with offline forms?

**Comment by signalpoint** [comment]
@metax45 There has been no progress on getting this into DrupalGap core. However, there has been much recent progress with a contributed DrupalGap module, called Local Forms: http://drupalgap.org/project/local_forms - I am currently building this for a project, so it will see much improvement over the next few weeks.

**Comment by chrisschaub** [comment]
This is cool stuff ...

http://aws.amazon.com/cognito/

and there are js/nodejs libraries to work with it.
- Chris

On Wed, Mar 18, 2015 at 4:06 PM, Tyler Frankenstein <
notifications@github.com> wrote:

> @metax45 https://github.com/metax45 There has been no progress on
> getting this into DrupalGap core. However, there has been much recent
> progress with a contributed DrupalGap module, called Local Forms:
> http://drupalgap.org/project/local_forms - I am currently building this
> for a project, so it will see much improvement over the next few weeks.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/signalpoint/DrupalGap/issues/258#issuecomment-83176921
> .

## 

Christopher Schaub
http://chris.schaub.com

**Comment by joeyshidab** [comment]
@signalpoint how is the progress with this?

**Comment by signalpoint** [comment]
@joeyshidab All progress has, and will be, reported here. Folks, I understand this issue is in high demand, we need someone to step forth and help establish the most feasible plan to implement this solution (or throw money at the probelm). All parties interested, come chat about it: www.drupalgap.org/irc

EDITED

**Comment by eduoda** [comment]
Hi, I'm studying the local_forms and I think it's too generic, it tries to deal with all sort of form which leads to several issues that I'm not sure we will find a simple way to deal with, e.g, page redirects.

Maybe we should restrict the local_forms to requests that are made to a Drupal services endpoint. I made a fork, so I can test this idea.

If I'm right, then probably this feature should be part of jDrupal and not DrupalGap.

Oda

**Comment by signalpoint** [comment]
@eduoda Thanks for chiming in. Yes, the local_forms module is definitely generic, and will not be able to handle all situations.

That's an interesting idea that this offline/sync caching layer should/could be up in jDrupal, very interesting indeed.

**Comment by eduoda** [comment]
Please, check it out: https://github.com/eduoda/local_forms
I also had to make to change jDrupal a little bit: https://github.com/eduoda/jDrupal

I think that this kind of offline behavior can not be transparent to other contrib modules. They should use hooks or callback to interact with local_forms for deal with redirects, reset forms, etc.

**Comment by eduoda** [comment]
btw, I think that the way the drupalgap.online is updated in my version of local_forms is a good idea. I'm considering to write a module that automatically gets you out of that "failed connection" page when you get connected.

**Comment by signalpoint** [comment]
@eduoda This is some great stuff you've implemented. Have you tried running the tests.html file to see if all the unit tests still pass in jDrupal with these changes. I now definitely see the advantage to your approach, so that custom forms can implement their own handlers, and we can just focus on the core forms.

@rjsteinert, I just wanted to ping you in on this thread, and to hopefully hear your thoughts about a local_forms layer in jDrupal to help with online/offline CRUD syncs.

**Comment by eduoda** [comment]
Hi Frank! I'm glad to help you with the awesome job you have done!

I've just commit another version of jDrupal that has this feature in core. I think it's pretty simple and backward compatible: https://github.com/eduoda/jDrupal
I also made a slight modification to webform: https://github.com/eduoda/webform

Sorry, I was not aware about the tests.html, I'll run the tests as soon as I can.

**Comment by eduoda** [comment]
It seems fine:

Tests completed in 43004 milliseconds.
60 assertions of 60 passed, 0 failed.

**Comment by eduoda** [comment]
Hi @signalpoint 
Did you review the changes I made to jDrupal? I'm planning to work on this issue next week and it would be nice to know if I'm in the right direction.

**Comment by signalpoint** [comment]
@eduoda I'm glad the tests passed, very cool. I've looked at all the code, and everything looks good. The only part I don't understand is the usage of `tries`: https://github.com/eduoda/jDrupal/blob/oda/src/services/services.js#L21 - Will you please elaborate on that?

I'm going to make a new branch over here and pull in your changes to start writing some offline tests on node CRUD (for starters). What is your next plan?

**Comment by signalpoint** [comment]
@eduoda Here's the new branch of jDrupal with your changes merged in, and some improvements from my side: https://github.com/easystreet3/jDrupal/tree/7.x-1.x-offline

**Comment by eduoda** [comment]
@signalpoint `tries` keep track of how many times we've tried to make the request. Since services_preprocess hook modifies the request, I think it should be called just in the first try. Also, other hooks may need to be aware of this information.

**Comment by rjcorwin** [comment]
Has anyone given [PouchDB.js](http://pouchdb.com/) + [Relaxed module](https://www.drupal.org/project/relaxed) a try? PouchDB is an in-browser database that you can map reduce query, go offline and sync later, and generally feel pretty good that it's a solid codebase based around the proven http://replication.io protocol. I've used PouchDB quite a bit with CouchDB, the Relaxed module apparently gives Drupal ability to talk the replication.io protocol which means you can sync content between Drupal and PouchDB. Even if the API that Relaxed offers isn't used, PouchDB is great way to store versioned content.

**Comment by signalpoint** [comment]
@rjsteinert I haven't tried either of them, my gut feeling is I wouldn't want (another) dependency for jDrupal, so I'm hoping we can accomplish this without extra tools.

**Comment by definitesols** [comment]
@signalpoint cai I use https://github.com/easystreet3/jDrupal/tree/7.x-1.x-offline in my drupalgap app to build an offiline form submission module.

**Comment by signalpoint** [comment]
@definitesols it isn't ready for production, but anyone interested in trying it is welcome to. It still needs development work, if I recall correctly.

**Comment by afidegnum** [comment]
What's the final conclusion on this project? how do we save the form offline?

**Comment by nevkatz** [comment]
I all,
I just was able to successfully use a test app I built DrupalGap on a device, and it was able to pull in data from my drupal site - a momentous occasion! I came across this thread as I was looking at options for offline support, as we are working on an app for areas with low connectivity.

While I don't need a form to work offline, I do need my app do display node content when it is offline (assuming it has been online earlier to grab it from the Drupal database). I gather that this is likely not built into DrupalGap yet, but from reviewing the posts, it seems as if I should be able to achieve this by including localStorage functionality in my app. Is this correct? Is there a way for the app to grab audio, video, and images from a drupal site and have it work locally, or should those assets be pre-packaged in the app?

**Comment by signalpoint** [comment]
@nevkatz Correct. This feature hasn't been built into DrupalGap yet. Attempts have been made with the Local Forms module. No matter what developers can use local storage to do many things.

I've had success getting offline functionality into production apps by using a combination of `deviceready`, `device_connection`, `device_offline` hooks in a custom DrupalGap module, and with the various services pre/post process hooks: http://api.drupalgap.org/7/

jDrupal also has some very basic local storage entity caching built in, which can be helpful at times.

As for audio/video etc, I wouldn't recommend pre-packaging that with the app, but instead dynamically download/cache/expire assets like that when connections to Drupal are availalbe.

**Comment by nevkatz** [comment]
Thanks so much for your quick reply. I will look into jDrupal and will see if I can implement one or more of those hooks you mentioned to make local storage happen. Looks like it's possible to build offline support on top of DrupalGap and it should be an exciting endeavor figuring it out. Very much appreciated!

**Comment by ghost** [comment]
@signalpoint do you have an example module using deviceready, device_connect, device_offline? I'm looking to access simple nodes on- & offline... thx!

**Comment by signalpoint** [comment]
@littlehkranch The facebook.js module for DrupalGap has an example of `hook_deviceready`, otherwise the API has examples of each: http://api.drupalgap.org/7/

**Comment by nevkatz** [comment]
Awesome Tyler, thanks so much for the tip. I am on the verge of getting what I need working. I have figured out local storage and am just trying to get my device to detect whether there is a change in its connectivity and then either go to its online or offline page depending on what the change is. Sounds like facebook.js and the API might help with this. I'm going to try coding something up shortly for this, maybe using a function that gets called at regular intervals.

**Comment by signalpoint** [comment]
The [drupalgap_has_connection](http://api.drupalgap.org/7/global.html#drupalgap_has_connection) function is also helpful here.

**Comment by nevkatz** [comment]
Hi Tyler, 

Thanks so much for your help on getting offline functionality working. What is the best method for replicating a view in offline mode? I have been able to recreate a list of node links without a view using some javascript, but I'm not sure if there's a way of actually using the views call with a modified path (or somehow making the existing path work.) 

Best,

Nevin

**Comment by manxon** [comment]
Hi nevkatz, 

Great you made it to work offline, can you please give us some insights of the process?

Thank you

Julian

**Comment by nevkatz** [comment]
Hi Julian, 

Sure. My goal is to grab the JSON files for views and nodes that are made possible by Drupal Services. I created a view for JSON content, which exists as a JSON file with a set URL. When the app comes online, it accesses that view and iterates through all the node data. For each node, it then accesses the node's JSON file. While searching through the JSON, it goes and looks for the various assets — namely the image and audio assets — and downloads those to the device. 

When the device goes offline, I have a function override for the system_offline_page that in turn calls functions needed for retrieving the stored JSON files, and then the related assets. Once it retrieves the JSON files data and has a path to the assets for a given node, it is ready to populate a page. 

Hope that helps!

Nevin

**Comment by gireeshscs** [comment]
Hi there,

I am a drupalgap developer, i saw you job post in upwork. I cant apply for your job becuase my profile not satisfying your JD. I have not read all the details, but i think you need to work offline in the absense of internet connectvity. I think we can acheive this by using GCM push notification plugin. 

Gireesh

**Comment by Boby** [comment]
following up this, has this ever been done? anyone succeeded using drupalgap offline storage + sync to drupal system?
